File: ch4/listing4.4.R
pkg <- c("dfoptim", "ggplot2", "data.table")
sapply(pkg, require, character.only = TRUE)
wk <- ifelse(.Platform$OS.type =="windows",
             shortPathName("E:/PPT/ComputationalModelling/ch4/"),
             "/media/yslin/Tui/ComputationalModelling/ch4/")
setwd(wk)
rm(list = ls())
## Maximum likelihood estimation
source("GCMprednoisy.R")


## Data ------------------------------------------------------------------
## Nosofsky, R. M., (1991). Tests of an exemplar model for relating perceptual
## classification and recognition memory. Journal of Experimental Psychology:
## Huaman Perception and Performance, 17, 3-27
## 
## discrepancy for power forgetting function
## 
## Obtain predicted probabilitie, Pa and Pb
## There were 2 responses per face from 40 participants
## Assumption 1: the matches between exemplars and the categorisation stimlus
## are noisy.
## Assumption 2: four weights for each dimension. w1 + w2 + w3 + w4 = 1
## Summed similarity means there are a similarity score for each feature of a
## face stimlus. The subject worked out a similarity score feature-after-feature.
## So in the process of matching a stimulus to an exemplar, the subject
## will computer a summed similarity score.

## The data were processed across responses and participants
## There were 80 subjects in the classification experiment and 58 subjects
## in the similarity-ratings study. 10 subjects in the similarity study
## were not included in the analyses. left 48 for analyses
N <- 2*40  ## assuming two responses per face from 40 learners

## Table A1:
## Physical specifications and multidimensional scaling (MDS) coordinates for 
## the 34 faces. 
## Eye heights, eye separation, nose length and mouth height
stim <- as.matrix(read.csv("faceStim.csv", header = FALSE))
## stim <- as.matrix(read.table("faceStim.csv", sep = ","))
colnames(stim) <- c("eh", "es", "nl", "mh")
head(stim)
str(stim)

exemplars <- list(a = stim[1:5, ], b = stim[6:10, ])
## exemplars

## Table 1. Only experiment 1A, learners only, observed
data <- scan(file = "facesDataLearners.txt")
data <- ceiling(data*N)
data
## w1 <- .25
## w2 <- .25
## w3 <- .25
bestfit <- 1e4

for(w1 in c(.25, .5, .75)) {
    for(w2 in c(.25, .5, .75)) {
        for(w3 in c(.25, .5, .75)) {
            
            fitres <- nmkb(par = c(1, w1, w2, w3, 1, .2),
                           fn = function(theta) GCMutil(theta, stim, exemplars, data, N, FALSE),
                           lower = c(0,  0, 0, 0,   0, -5),
                           upper = c(10, 1, 1, 1,  10,  5),
                           control = list(trace = 0))
            ## fitres$value is the deviance

            ## retain the smallest deviance
            if (fitres$value < bestfit) {
                bestres <- fitres
                bestfit <- fitres$value
            }
        }
    }
}

round(bestres$par, 2)
## c,   w1,  w2,  w3,   sigma, b
## 2.55 0.37 0.01 0.98  1.15   0.08

theta <- bestres$par
w <- theta[2]
w[2] <- (1 - w[1])*theta[3]
w[3] <- (1 - sum(w[1:2]))*theta[4]
w[4] <- (1 - sum(w[1:3]))
round(w, 2)
## eye height and nose length got more attention than
## eye separation and mouth height
##      eh   es   nl   mh
## c    w1   w2   w3   w4   sigma b
## 2.55 0.37 0.01 0.61 0.01 1.15  0.08
stim
## return predicted probability
preds <- GCMutil(bestres$par, stim, exemplars, data, N, TRUE)
round(preds, 2) - data

pdf(file = "GCMfits.pdf", width = 5, height = 5)
plot(preds, data/N, xlab = "Data", ylab = "Predictions")
dev.off()

bestres$par

File: ch4/listing4.4.py
## Yi-Shin Lin, 16-11-2021
import numpy as np
from scipy.stats import binom
from scipy.stats import norm
from scipy.optimize import minimize
import pandas as pd
import math

## Obtain current predictions and compute discrepancy
def Euclidean(x, y, w):
    ## Euclidean(stim[:1].to_numpy(), stim[:2].to_numpy(), w)
    term0 = w*np.power(x - y, 2)
    out = np.sqrt( np.sum(term0) )
    return out

def GCMprednoisy(probe, exemplars, c, w, sigma, b):
    '''
    Calculate GCM likelihood of categorising a face stimulus, `probe` into category `A` out of N total responses, given parameter c, w, and b.

    Arguments:
	    probe: a numpy array 
	    exemplars: a two-element dictionary
	    c: a number
	    w: a numpy array
            sigma: a number
	    b: a number
    Return:
            a probability of categorising stimulus, `probe` into category A
    Example:
        res = GCMpred(stim[:1].to_numpy(), exemplars, c, w, sigma, b)
    '''
    nexem = len(exemplars)
    sim_score = np.empty(nexem)
    sim_score[:] = np.nan
    
    for i, key in enumerate(exemplars.keys()):
        ex = exemplars[key]
        nobs = len(ex)    ## number of observations in an exemplar
        
        tmp = np.empty(nobs)
        tmp[:] = np.nan
        for j in range(0, nobs):
            tmp[j] = Euclidean(ex.iloc[j].to_numpy(), probe, w)

        sim_score[i] = np.sum(np.exp(-c*tmp))

    ## Whether the difference of similarity score exceeds a threshold b
    out = norm.cdf(sim_score[0] - sim_score[1] - b, loc = 0, scale = sigma)
    return out

def GCMutil(theta):
    ## , stim, exemplars, data, N):
    ## theta = np.array([1, .25, .25, .25, 1, .2])
    ## nDat = len(data)
    ## theta = init_guess
    dev = np.empty(nDat)
    pre = np.empty(nDat)
    dev[:] = np.nan
    pre[:] = np.nan
    w = np.empty(4)
    w[:] = np.nan

    c = theta[0]
    w[0] = theta[1]
    w[1] = (1 - w[0]) * theta[2]
    w[2] = (1 - np.sum(w[:2])) * theta[3]
    w[3] = (1 - np.sum(w[:3]))
    sigma = theta[4]
    b = theta[5]

    for i in range(0, nDat):
        p = GCMprednoisy(stim.iloc[i].to_numpy(), exemplars, c, w, sigma, b)
        dev[i] = -2*np.log(binom.pmf(k = data[i], n = N, p = p))

    out = np.sum(dev)
    return out
    

N = 2*40 ## assuming two responses per face from 40 learners
stim = pd.read_csv('faceStim.csv', header = None)
exemplars = {
    "A": stim[:5],
    "B": stim[5:10]
}


tmp = pd.read_table("facesDataLearners.txt", header = None)
data = np.ceil(tmp.to_numpy()*N)
nDat = len(data)
bestfit = 1e4

w1 = np.array([.25, .5, .75])
w2 = np.array([.25, .5, .75])
w3 = np.array([.25, .5, .75])
bnds = ((0, 10), (0, 1), (0, 1), (0, 1), (0, 10), (-5, 5))

for i in w1:
     for j in w2:
         for k in w3:
             init_guess = np.array([1, i, j, k, 1, .2])
             res = minimize(fun = GCMutil, x0 = init_guess, method='L-BFGS-B', bounds = bnds)
             # res = minimize(fun = GCMutil, x0 = init_guess, method='Nelder-Mead', options={'xatol': 1e-8, 'disp': True})
             if(res.fun < bestfit):
                     bestres = res
                     bestfit = res.fun

bestres.x
bestres.fun
## 153.10190314638706
## array([2.54562316, 0.37134223, 0.00864732, 0.97985185, 1.15405855,
##        0.07934023])
